Adding badge to legacy admin menu to indicate remaining onboarding tasks

joelclimbsthings opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adding a badge to the home-screen item on the admin menu to indicate how many setup tasks remain. 

_Notes:_
- I also added the badge to the new woo navigation Home item, although I noticed some buggy behavior. Given the state of the woo navigation project I wasn't sure whether to pursue it further in the context of this PR, so I left it as is for now.
- The admin menu is only updated on a page refresh, so if a task completes with this it won't updated the badge. Updating this on the legacy admin menu will likely be somewhat hacky, but I'm glad to do this if we deem it necessary. 

![image](https://user-images.githubusercontent.com/444632/163244183-66822cfc-0f7b-47af-b0c3-6de266fcb47a.png)


Closes #32160  .

### How to test the changes in this Pull Request:

1. Checkout this branch on a fresh site.
2. Note that this should work with either task list experiment or without entirely.
3. Notice that no badge is displayed before completing any tasks (Unless one is auto-completed, which is true for at least one of the experiments)
4. Complete the setup wizard, and notice that the badge is now displayed indicated remaining tasks
5. Complete a couple other tasks, notice that the badge is updated properly. (don't complete all of them)
6. Hide the task list, and notice that the badge disappears
7. Display the task list again through Help -> Setup Wizard -> Task List -> Enable
8. Notice badge reappears
9. Complete the setup task list.
10. Notice the badge now disappears again.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [x] Have you created a changelog file by running `pnpm nx affected --target=changelog`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Commit:
Adding badge to legacy menu for onboarding tasks remaining

Commit:
Adding basic support for tasks badge to woo navigation

Commit:
Adding changelog

Commit:
Correcting spacing manually since autoformatter isn't working

joelclimbsthings commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskLists.php:
```diff
@@ -375,6 +376,55 @@ function ( $all, $curr ) {
+			return;
+		}
+
+		foreach ( $submenu['woocommerce'] as $key => $menu_item ) {
```
This is using basically the same logic and styles as the preexisting orders badge, but I'm receptive to a more elegant way if one exists.

octaedro commented:
Good job @joelclimbsthings! The code looks good and it's working as expected but I'm not sure about the current behavior.

There are a few tasks that don't refresh the page after finishing them and the badge doesn't update, creating some confusion.

![screenshot-strict-finch jurassic ninja-2022 04 21-14_37_58](https://user-images.githubusercontent.com/1314156/164518825-b8840041-8910-4257-89d3-a729e6e0e1c1.png)

Is it possible to update the remaining tasks badge on the fly?
Another option that I can imagine is to force a page refresh after finishing each step. I would add a check to confirm that the experiment is turned on to narrow down that behavior just to the experiment.

joelclimbsthings commented:
Thanks for the review @octaedro !
> There are a few tasks that don't refresh the page after finishing them and the badge doesn't update, creating some confusion.

Yeah, I actually mentioned this in my "_Notes_" above. We can definitely do this, although it would just be altering the markup via JS, so feels a bit hacky. That said I still may be in favor of it. 

Also, we'd want to ensure that's it's _only_ executed when a task doesn't result in a page refresh, and I'm not sure offhand if there's a good preexisting place to do that, or if we'd have to do something tricky. 

> Just in case, referring to step 3 in the testing instructions; if I skip the OBW (without finishing any tasks) I see the badge displaying the remaining tasks. Is that expected?

Actually not sure on this one. The [original issue](https://github.com/woocommerce/woocommerce/issues/32160) doesn't mention that as an exception, but we'd have to ask @jarekmorawski to clarify.

octaedro commented:
Thank you @joelclimbsthings for your answer.

> Yeah, I actually mentioned this in my "Notes" above. We can definitely do this, although it would just be altering the markup via JS, so feels a bit hacky. That said I still may be in favor of it.

Since this is something with high visibility, I would be cautious about possible inconsistencies because they could be seen as bugs by the user that doesn't refresh the page. Is there a chance for us to improve that?

// CCing @pmcpinto and @jarekmorawski , to have more insights about the expected behavior here.

Commit:
Altering badge logic to function based on experiments instead of feature flag

Commit:
Updating badge for tasks that complete with no page refresh

joelclimbsthings commented:
Thanks @octaedro , I've gone ahead and tentatively added the dynamic update in 48bccc1fbe0b9d491510f00d0b4df8e7c4d03d97. I've also updated the feature flag logic to work with the setup task list experiments, and rebased the branch on the current state of `trunk`.

pmcpinto commented:
> There are a few tasks that don't refresh the page after finishing them and the badge doesn't update, creating some confusion.

Considering this, I wonder I we should show just "1" in the badge instead of the number of incomplete tasks. @jarekmorawski @verofasulo what do you think?

jarekmorawski commented:
> Considering this, I wonder I we should show just "1" in the badge instead of the number of incomplete tasks. @jarekmorawski @verofasulo what do you think?

"1" can be misleading since it's not entirely clear what it'd mean. How about an asterisk or a simple red dot? Anything to indicate there are things in there that require people's attention. 🤔

<img width="473" alt="image" src="https://user-images.githubusercontent.com/79307566/165276704-a3b58ae1-77f9-4ce6-b6f2-bd5335b77f2d.png">

octaedro commented on plugins/woocommerce-admin/client/tasks/task.tsx:
```diff
@@ -25,6 +25,19 @@ export const Task: React.FC< TaskProps > = ( { query, task } ) => {
+			return;
+		}
+
+		const currentBadgeCount = Number( badgeElement.innerText );
```
Can we add a check here to prevent showing `0`?

It might be something like this:
```
if ( currentBadgeCount === 1 ) {
	badgeElement.remove();
}
```

joelclimbsthings replied:
Good thought, done in https://github.com/woocommerce/woocommerce/pull/32611/commits/d3ff719428fe87649166064e351249131d84c74a

octaedro commented on plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskLists.php:
```diff
@@ -461,6 +462,58 @@ function ( $all, $curr ) {
+	 */
+	public static function setup_tasks_remaining () {
+
+		$active_list = self::is_experiment_treatment( 'woocommerce_tasklist_setup_experiment_1' ) ? 'setup_experiment_1' : self::is_experiment_treatment( 'woocommerce_tasklist_setup_experiment_2' ) ? 'setup_experiment_1' : null;
```
This works well on my end but on fresh installs (using JN) I'm seeing a couple of PHP warnings thrown from this line:

```
Cannot modify header information - headers already sent by (output started at /public/wp-content/plugins/woocommerce/src/Admin/Features/OnboardingTasks/TaskLists.php:472)
```
and
```

Unparenthesized `a ? b : c ? d : e` is deprecated. Use either `(a ? b : c) ? d : e` or `a ? b : (c ? d : e)`
```

joelclimbsthings replied:
Ah, look like I had my debugging flag disabled and didn't see this. Thanks for the heads up 👍🏻

octaedro reviewed (commented):
Thank you @joelclimbsthings for adding the dynamic update. The changes look good, I just left a couple of small comments.

Commit:
Reworking condition to allow reminder badge with corret list for all experiment states and control

Commit:
Hiding badge when remaining tasks is now zero

Commit:
Enabling reminder bar for both experiments and control

joelclimbsthings commented:
Hey @octaedro , I've resolved the issues you brought up, and also took the liberty of tweaking the condition controlling the reminder bar since it was [out of date](p1650967051612209/1650923410.541359-slack-C9K5T0XSP).

jacob-sewell reviewed (approved):
The only hiccup I had was that hiding the task list didn't result in the badge disappearing until after a refresh.

joelclimbsthings commented:
Thanks!

> The only hiccup I had was that hiding the task list didn't result in the badge disappearing until after a refresh.

Nice catch. Since the remedy would be fairly inelegant and hacky, and it's a fairly rare event, I'm opting for leaving it as is for now. We can revisit if needed.

tammullen commented:
Just noting this PR was merged with a failing k6 test due to the badge addition causing the k6 test assertion (which used the home menu item) to fail.

PR to update the assertion here https://github.com/woocommerce/woocommerce/pull/33044
